File: Airwallex/Core/Sources/AWXPaymentConsent.m
//
//  AWXPaymentConsent.m
//  Airwallex
//
//  Created by Victor Zhu on 2021/3/25.
//  Copyright © 2021 Airwallex. All rights reserved.
//

#import "AWXPaymentConsent.h"
#import "AWXPaymentMethod.h"

@implementation AWXPaymentConsent

+ (id)decodeFromJSON:(NSDictionary *)json
{
    AWXPaymentConsent *intent = [AWXPaymentConsent new];
    intent.Id = json[@"id"];
    intent.requestId = json[@"request_id"];
    intent.customerId = json[@"customer_id"];
    intent.status     = json[@"status"];
    intent.nextTriggeredBy = json[@"next_triggered_by"];
    intent.merchantTriggerReason = json[@"merchant_trigger_reason"];
    intent.requiresCVC = [json[@"requires_cvc"] boolValue];
    NSDictionary *paymentMethod = json[@"payment_method"];
    if (paymentMethod) {
        intent.paymentMethod = [AWXPaymentMethod decodeFromJSON:paymentMethod];
    }
    intent.createdAt = json[@"created_at"];
    intent.updatedAt = json[@"updated_at"];
    intent.clientSecret = json[@"client_secret"];
    return intent;
}

@end

File: Airwallex/Core/Sources/Internal/Models/AWXPaymentConsent.swift
//
//  AWXPaymentConsent.swift
//  Core
//
//  Created by Tony He (CTR) on 2024/7/22.
//  Copyright © 2024 Airwallex. All rights reserved.
//

import Foundation

/// `AWXPaymentConsent` includes the info of payment consent.
@objcMembers
@objc
public class AWXPaymentConsent: NSObject, Codable {
    /**
     Consent ID.
     */
    public let id: String?

    /**
     Request ID.
     */
    public let requestId: String?

    /**
     Customer ID.
     */
    public let customerId: String?

    /**
     Consent status.
     */
    public let status: String?

    /**
     Payment method.
     */
    public let paymentMethod: AWXPaymentMethod?

    /**
     Next trigger By type.
     */
    public let nextTriggeredBy: String?

    /**
     Merchant trigger reason
     */
    public let merchantTriggerReason: String?

    /**
     Whether it requires CVC.
     */
    public let requiresCVC: Bool = false

    /**
     Created at date.
     */
    public let createdAt: String?

    /**
     Updated at date.
     */
    public let updatedAt: String?

    /**
     Client secret.
     */
    public let clientSecret: String?

    enum CodingKeys: String, CodingKey {
        case id
        case requestId = "request_id"
        case customerId = "customer_id"
        case status
        case paymentMethod = "payment_method"
        case nextTriggeredBy = "next_triggered_by"
        case merchantTriggerReason = "merchant_trigger_reason"
        case requiresCVC = "requires_cvc"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case clientSecret = "client_secret"
    }

    public init(id: String?, requestId: String?, customerId: String?, status: String?, paymentMethod: AWXPaymentMethod?, nextTriggeredBy: String?, merchantTriggerReason: String?, createdAt: String?, updatedAt: String?, clientSecret: String?) {
        self.id = id
        self.requestId = requestId
        self.customerId = customerId
        self.status = status
        self.paymentMethod = paymentMethod
        self.nextTriggeredBy = nextTriggeredBy
        self.merchantTriggerReason = merchantTriggerReason
        self.createdAt = createdAt
        self.updatedAt = updatedAt
        self.clientSecret = clientSecret
    }

    public static func decodeFromJSON(_ dic: [String: Any]) -> AWXPaymentConsent {
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: dic, options: [])
            let result = try JSONDecoder().decode(AWXPaymentConsent.self, from: jsonData)

            return result
        } catch {
            return AWXPaymentConsent(id: nil, requestId: nil, customerId: nil, status: nil, paymentMethod: nil, nextTriggeredBy: nil, merchantTriggerReason: nil, createdAt: nil, updatedAt: nil, clientSecret: nil)
        }
    }
}
